qiskit/transpiler/basepasses.py

```diff
@@ -11,24 +11,92 @@
+from qiskit.converters import circuit_to_dag, dag_to_circuit
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.passmanager.base_optimization_tasks import GenericPass, OptimizerTask
+from qiskit.passmanager.propertyset import PropertySet, PassState, RunState
 
+from .exceptions import TranspilerError
+from .fencedobjs import FencedDAGCircuit
 from .layout import TranspileLayout
 
 
-class BasePass(GenericPass):
+class MetaPass(type):
+    """Metaclass for transpiler passes.
+
+    Enforces the creation of some fields in the pass while allowing passes to
+    override ``__init__``.
+    """
+
+    # Drop this functionality in the future.
+    # This metaclass provides a pass equivalence evaluation based on the constructor arguments.
+    # This implicit fake-hash based equivalence is fragile, and the pass developer must
+    # explicitly implement equivalence check logic for each pass if necessary.
+    # Currently, this metaclass is just here for backward compatibility, because
+    # circuit pass manager has a functionality to avoid multiple execution of the
+    # same pass (even though they are scheduled so). This is managed by the valid_passes set,
+    # and executed passes are added to this collection to avoid future execution.
+    # Dropping this metaclass causes many unittest failures and thus this is
+    # considered as a breaking API change.
+    # For example, test.python.transpiler.test_pass_scheduler.TestLogPasses.test_passes_in_linear
+
+    def __call__(cls, *args, **kwargs):
+        pass_instance = type.__call__(cls, *args, **kwargs)
+        pass_instance._hash = hash(MetaPass._freeze_init_parameters(cls, args, kwargs))
+        return pass_instance
+
+    @staticmethod
+    def _freeze_init_parameters(class_, args, kwargs):
+        self_guard = object()
+        init_signature = signature(class_.__init__)
+        bound_signature = init_signature.bind(self_guard, *args, **kwargs)
+        arguments = [("class_.__name__", class_.__name__)]
+        for name, value in bound_signature.arguments.items():
+            if value == self_guard:
+                continue
+            if isinstance(value, Hashable):
+                arguments.append((name, type(value), value))
+            else:
+                arguments.append((name, type(value), repr(value)))
+        return frozenset(arguments)
+
+    def mro(cls) -> list[type]:
+        # Avoid metaclass conflict.
+        mro_cls = super().mro()
+        if GenericPass not in mro_cls:
+            mro_cls.pop()
+            mro_cls += [GenericPass, OptimizerTask, object]
+        return mro_cls
+
+
+class BasePass(metaclass=MetaPass):
     """Base class for transpiler passes."""
 
+    def __init__(self):
+        super().__init__()
+        self.preserves: Iterable[GenericPass] = []
+        self._hash = hash(None)
+
+    def __hash__(self):
+        return self._hash
+
+    def __eq__(self, other):
+        return hash(self) == hash(other)
```

Comment:
Yes I totally agree. Probably directly comparing init args without hashing is a good option, but this still cannot cope with class name conflict issue.

Reply:
I see that the code I highlighted here is actually pre-existing, so let's leave it for the purposes of this PR, but we should revisit it in the future because it's incorrect.